### PR TITLE
8295278: Add parallel class loading tests

### PR DESCRIPTION
**Pull request by coleenp**
I moved the ParallelLoad test to a subdirectory, and created a new one called CallLoadClass that calls loadClass directly and then waits on the ClassLoader object, while another thread calls forName() to go through the SystemDictionary resolution.

I also made dedicated tests out of the bugs reported in
[JDK-4699981](https://bugs.openjdk.org/browse/JDK-4699981) ClassCircularityError thrown without reason during class loading and
[JDK-4670071](https://bugs.openjdk.org/browse/JDK-4670071) java.lang.ClassLoader.loadClassInternal(String) is too restrictive

Tested with tier1 on Oracle platforms. These tests are run in tier1.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295278](https://bugs.openjdk.org/browse/JDK-8295278): Add parallel class loading tests

### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [abab0205](https://git.openjdk.org/jdk/pull/10587/files/abab0205643f353771fdd0c84bc82ae950af3430)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [796c7b45](https://git.openjdk.org/jdk/pull/10587/files/796c7b45d97821935860af26f1b8d63f4cb25d8f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10587/head:pull/10587` \
`$ git checkout pull/10587`

Update a local copy of the PR: \
`$ git checkout pull/10587` \
`$ git pull https://git.openjdk.org/jdk pull/10587/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10587`

View PR using the GUI difftool: \
`$ git pr show -t 10587`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10587.diff">https://git.openjdk.org/jdk/pull/10587.diff</a>

</details>

### CODE REVIEW

**Review by iwanowww — APPROVED** [review]
Looks good.

File: test/hotspot/jtreg/runtime/ParallelLoad/CallLoadClass/ClassLoadingThread.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            success = false; // Should have thrown LinkageError
+        } catch (Throwable e) {
+            // If you call loadClass directly, this will result in LinkageError
+            ThreadPrint.println("Exception is caught: " + e);
+            e.printStackTrace();
+            if (e instanceof LinkageError) {
+                success = true;
+            } else {
+                success = true;
+            }
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/ParallelLoad/CallLoadClass/ClassLoadingThread.java:68]: Should `success` be set to `false` in this case?

**Review by calvinccheung — COMMENTED** [review]
I looked at mostly the tests under CallLoadClass and have one comment. Thanks.

File: test/hotspot/jtreg/runtime/ParallelLoad/CallLoadClass/ClassLoadingThread.java (hunk 1 above)
Comment (coleenp) [reply test/hotspot/jtreg/runtime/ParallelLoad/CallLoadClass/ClassLoadingThread.java:68]: Yes, you're right, it should be false.  We expect LinkageError in this case.

**Comment by coleenp** [comment]
Thanks for reviewing, Vladimir and Calvin.

**Review by calvinccheung — APPROVED** [review]
Updates look good. Thanks.

File: test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/test-classes/A.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+// A extends B in other Class loader.
+public class A  extends B {
```
Comment (turbanoff) [inline test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/test-classes/A.java:25]:
```suggestion
public class A extends B {
```

**Comment by coleenp** [comment]
Thank you Vladimir and Calvin for the reviews, and @turbanoff for the suggestion to remove the unneeded space.
/integrate

**Pull request closed without merging**